osm2streets/src/geometry/algorithm.rs

```diff
@@ -80,224 +76,135 @@ pub fn intersection_polygon(
+        orig_centers.insert(*id, road.center_line.clone());
     }
 
-    // Intersect every road's boundary lines with all the other lines. Only side effect here is to
-    // populate new_road_centers.
-    let mut new_road_centers: BTreeMap<RoadID, PolyLine> = BTreeMap::new();
-    // TODO If Results has a BTreeMap too, we could just fill this out as we go
-    for (r1, pl1) in &road_lines {
-        // road_center ends at the intersection.
-        let road_center = if roads[r1].dst_i == i {
-            roads[r1].center_line.clone()
-        } else {
-            roads[r1].center_line.reversed()
-        };
+    // Look at every adjacent pair of edges
+    let mut edges = RoadEdge::calculate(sorted_roads.iter().collect(), results.intersection_id);
+    edges.push(edges[0].clone());
 
-        // Always trim back a minimum amount, if possible.
-        let mut shortest_center = road_center.clone();
+    for pair in edges.windows(2) {
+        let one = &pair[0];
+        let two = &pair[1];
 
-        for (r2, pl2) in &road_lines {
-            if r1 == r2 {
-                continue;
-            }
-
-            // If two roads go between the same intersections, they'll likely hit at the wrong
-            // side. Just use the second half of the polyline to circumvent this. But sadly, doing
-            // this in general breaks other cases -- sometimes we want to find the collision
-            // farther away from the intersection in question.
-            let same_endpoints = {
-                let ii1 = roads[r1].src_i;
-                let ii2 = roads[r1].dst_i;
-                let ii3 = roads[r2].src_i;
-                let ii4 = roads[r2].dst_i;
-                (ii1 == ii3 && ii2 == ii4) || (ii1 == ii4 && ii2 == ii3)
-            };
-            let (use_pl1, use_pl2): (PolyLine, PolyLine) = if same_endpoints {
-                (pl1.second_half()?, pl2.second_half()?)
-            } else {
-                (pl1.clone(), pl2.clone())
-            };
+        // Only want corners between two roads
+        if one.road == two.road {
+            continue;
+        }
 
-            if use_pl1 == use_pl2 {
-                bail!(
-                    "{r1} and {r2} have overlapping segments. You likely need to fix OSM and make the \
-                     two ways meet at exactly one node."
-                );
-            }
+        // Look for where the two road edges collide, closest to the intersection.
+        if let Some((pt, _)) = one.pl.reversed().intersection(&two.pl.reversed()) {
+            // For both edges, project perpendicularly back to the original center, and trim back
+            // to that point.
+            for side in [one, two] {
+                if let Some((_, angle)) = side.pl.dist_along_of_point(pt) {
+                    let perp = InfiniteLine::from_pt_angle(pt, angle.rotate_degs(90.0));
+
+                    // Make the road center point away from the intersection, so we find the hit
```

Comment:
Lots of the complexity comes from this. Most of the time polylines just hit at one point and it doesn't matter. But we have to spend lots of effort here to find the hitpoint closest to the intersection we're operating on. There's probably an easier way to do this.